File: app/src/main/java/de/markusressel/typedpreferencesdemo/dagger/AppModule.java
package de.markusressel.typedpreferencesdemo.dagger;

import android.app.Application;
import android.content.Context;

import javax.inject.Singleton;

import dagger.Binds;
import dagger.Module;
import dagger.Provides;
import dagger.android.ContributesAndroidInjector;
import de.markusressel.typedpreferencesdemo.MainActivity;
import de.markusressel.typedpreferencesdemo.application.App;

/**
 * Created by Markus on 11.07.2017.
 */
@Module
public abstract class AppModule {

    @Binds
    abstract Application application(App application);

    @Provides
    @Singleton
    static Context provideContext(Application application) {
        return application;
    }

    @ContributesAndroidInjector
    abstract MainActivity mainActivity();

}

File: app/src/main/java/de/markusressel/typedpreferencesdemo/dagger/AppModule.kt
/*
 * Copyright (c) 2017 Markus Ressel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.markusressel.typedpreferencesdemo.dagger

import android.app.Application
import android.content.Context
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.android.ContributesAndroidInjector
import de.markusressel.typedpreferencesdemo.MainActivity
import de.markusressel.typedpreferencesdemo.application.App
import javax.inject.Singleton

/**
 * Created by Markus on 11.07.2017.
 */
@Module
abstract class AppModule {

    @Binds
    internal abstract fun application(application: App): Application

    @ContributesAndroidInjector
    internal abstract fun mainActivity(): MainActivity

    @Module
    companion object {

        @Provides
        @Singleton
        @JvmStatic
        internal fun provideContext(application: Application): Context {
            return application
        }
    }

}
